docs/src/extend/custom-rule-tutorial.md

```diff
@@ -0,0 +1,456 @@
+
+To learn more about custom rules and plugins refer to the following documentation:
+
+* [Custom Rules](custom-rules)
+* [Plugins](plugins)
+
+## Why Create a Custom Rule?
+
+Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, prevent a particular bug from recurring, or ensure compliance with a style guide.
+
+Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see if someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.
+
+## Prerequisites
+
+Before you begin, make sure you have the following installed in your development environment:
+
+* [Node.js](https://nodejs.org/en/download/)
+* [npm](https://www.npmjs.com/)
+
+This tutorial also assumes that you have a basic understanding of ESLint and ESLint rules.
+
+## The Custom Rule You Make
+
+The custom rule that you make in this tutorial requires that all `const` variables named `foo` are assigned the string `"bar"`. The rule is defined in the file `foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.
+
+You can run the custom rule on a file to report and fix any problems.
+
+For example, say you had the following `foo.js` file:
+
+```javascript
+// foo.js
+
+const foo = "baz123";
+```
+
+Running ESLint with the rule would fix the file to contain the following:
+
+```javascript
+// foo.js
+
+const foo = "bar";
+```
+
+## Step 1: Set up Project
+
+First, create a new project for your custom rule. Create a new directory, initiate a new npm project in it, and create a new file for the custom rule:
+
+```shell
+mkdir eslint-custom-rule-example # create directory
+cd eslint-custom-rule-example # enter the directory
+npm init -y # init new npm project
+touch enforce-foo-bar.js # create file foo-bar.js
+```
+
+## Step 2: Stub Out Rule File
+
+In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also, add a `meta` object with some basic information about the rule.
+
+All ESLint rules are objects that follow this structure.
+
+```javascript
+// enforce-foo-bar.js
+
+module.exports = {
+  meta: {
+     // TODO: add metadata
+    },
+    create(context) {
+        return {
+            // TODO: add callback function(s)
+        };
+    }
+};
+```
+
+## Step 3: Add Rule Metadata
+
+Before writing the rule, add some metadata to the rule object. ESLint uses this information when running the rule.
+
+Start by exporting a module with a `meta` object containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "fix," the description is "Can only assign 'bar' to const foo.", and the rule is fixable by modifying the code.
```

Comment:
Looks like the code and the description have fallen out of sync. Please double check throughout.

```suggestion
Start by exporting an object with a `meta` property containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "problem," the description is "Enforce that a variable named `foo` can only be assigned a value of 'bar'.", and the rule is fixable by modifying the code.
```